dips/dip-168.md

```diff
@@ -0,0 +1,114 @@
+4. **Dependency:** Dependency (but not strict sequentiality) must be able to be expressed between two transactions. e.g., an account should be able to guarantee that `txn2` will not be executed before `txn1` by some other means than waiting for `txn1` to be executed and committed before submitting `txn2`.
+5. **Space:** The representation of the CRSN must have a constant-size space footprint.
+6. **Opt-in:** Using a CRSN must be opt-in at the account level. An account must be able to opt-out of using a CRSN at any point. The system must be able to support accounts with normal sequence numbers and other accounts with CRSNs coexisting in the network.
+
+# CRSN Definition
+
+We define a CRSN as follows. A CRSN `CRSN(k)` is a stateful object published under an account `A` and parameterized by a bounded number `k`. `CRSN(k)` holds two pieces of data:
+
+* A `min_nonce: u64` field
+* A `slots: [bool; k]` field consisting of `k` boolean slots initialized to `false`.
+
+Each transaction sent from `A` is assigned a number or *sequence nonce* `i: u64`. This number coupled with the state `CRSN(k)` held under `A` defines the acceptance and rejection criteria for transactions.
+
+## Acceptance and Rejection Criteria
+
+The acceptance and rejection criteria for a pair `(i, CRSN(k))` where `i` is the sequence nonce assigned to a transaction sent by `A` and `CRSN(k)` is the CRSN state under `A` at the time of processing is as follows:
+
+1. If `i` is greater than or equal to `CRSN(k).min_nonce + k` the transaction    is discarded.
+2. If `i` is less than `CRSN(k).min_nonce` the transaction is discarded.
+3. If `i` is greater than or equal to `CRSN(k).min_nonce` and less than    `CRSN(k).min_nonce + k` then
+    1.  If the slot at position `i` in `CRSN(k).slots` is `false` then `CRSN(k).slots[i]` is set to `true` and the transaction is accepted. Subsequently, the [shifting algorithm](#shifting-algorithm) is applied to `CRSN(k)`. `A`'s sequence number is incremented.
+    2. If the slot at position `i` in `CRSN(k).slots` is `true` the transaction is rejected.
+
+## Shifting algorithm
+
+Let `j = max {m | CRSN(k).min_nonce <= e <= m < k . forall CRSN(k).slots[e] == true }` (i.e., the least index in `CRSN(k).slots` such that all slots less than it are set to `true`).  The following updates are then performed:
+
+* `CRSN(k).min_nonce = CRSN(k).min_nonce + j`
+* for all `j <= i < k, ``CRSN(k).slot[i - j] = CRSN(k).slot[i]` and, for all `k - j <= i < k, ``CRSN(k).slot[i] = false`.
+
+## Sequence number to CRSN transition
```

Comment:
Good question. Sam and I discussed this, and while an account could transition back, the account's sequence number  when transitioning from CRSN -> seqno would have to be set at `CRSN(k).min_nonce + k` to be safe. This is because the account could have processed transactions with sequence nonces anywhere in the range of `(CRSN(k).min_nonce, CRSN(k).min_nonce + k)` and if we didn't do this it could allow transaction replay for transactions in this window.

This however causes issues if we want to keep sequence numbers sequential, because this would cause a jump in the sequence number at the time of transition.

Reply:
Yeah, that seems to violate the "A.sequence_number == <number of transactions sent by A that have been processed>"